### PR TITLE
Enable and configure KSM according to memory pressure

### PR DESCRIPTION
**Pull request by jean-edouard**
**What this PR does / why we need it**:
Enabling KSM has a substantial impact on performance, and isn't useful when there's plenty of memory available.
This PR preserves and respects the existing API, but tells KSM to run only when there's less than 20% memory available.
There's also some logic to tweak a couple parameters depending on the amount of available memory and the amount of time it's been low.

**Which issue(s) this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close the issue(s) when PR gets merged)*:
Fixes #

**Special notes for your reviewer**:
- As stated in the code, the logic was inspired from [oVirt](https://github.com/oVirt/mom/blob/master/doc/ksm.rules)

**Checklist**

This checklist is not enforcing, but it's a reminder of items that could be relevant to every PR.
Approvers are expected to review this list.

- [ ] Design: A [design document](https://github.com/kubevirt/community/tree/main/design-proposals) was considered and is present (link) or not required
- [ ] PR: The PR description is expressive enough and will help future contributors
- [ ] Code: [Write code that humans can understand](https://en.wikiquote.org/wiki/Martin_Fowler#code-for-humans) and [Keep it simple](https://en.wikipedia.org/wiki/KISS_principle)
- [ ] Refactor: You have [left the code cleaner than you found it (Boy Scout Rule)](https://learning.oreilly.com/library/view/97-things-every/9780596809515/ch08.html)
- [ ] Upgrade: Impact of this change on upgrade flows was considered and addressed if required
- [ ] Testing: New code requires [new unit tests](https://github.com/kubevirt/kubevirt/blob/main/docs/reviewer-guide.md#when-is-a-pr-good-enough). New features and bug fixes require at least on e2e test
- [ ] Documentation: A [user-guide update](https://github.com/kubevirt/user-guide/) was considered and is present (link) or not required. You want a user-guide update if it's a user facing feature / API change.
- [ ] Community: Announcement to [kubevirt-dev](https://groups.google.com/g/kubevirt-dev/) was considered

**Release note**:

```release-note
Dynamic KSM enabling and configuration
```

### CODE REVIEW

**Review by fossedihelm — COMMENTED** [review]
Thanks @jean-edouard! Wonderful work! I'm wondering whether it is worth adding a functest!
/lgtm

**Comment by fossedihelm** [comment]
@jean-edouard can you address the compute failures? They are ksm related. Thanks

**Comment by fossedihelm** [comment]
Thanks @jean-edouard! Can I ask you to add the functests deletion reasoning in the commit description too? In this way future us will be able to understand it from the git log. Thank you

**Comment by jean-edouard** [comment]
@fossedihelm I brought back the functional tests with a few debug annotations, PTAL!

File: tests/infrastructure/virt-handler.go [hunk 1]
```diff
@@ -47,36 +49,64 @@ var _ = DescribeInfra("virt-handler", func() {
...
 			Fail("There isn't any node with KSM available")
 		}
+
+		forceMemoryPressureOnNodes(nodesToEnableKSM)
+
 		originalKubeVirt = util.GetCurrentKv(virtClient)
 	})
 
 	AfterEach(func() {
+		restoreNodes(nodesToEnableKSM)
```
Comment (fossedihelm) [inline tests/infrastructure/virt-handler.go:109]: So sad global `CleanNodes` function does not clean the annotations

File: tests/infrastructure/infrastructure.go [hunk 2]
```diff
@@ -26,5 +26,5 @@ import (
 )
 
 func DescribeInfra(text string, args ...interface{}) bool {
-	return Describe("[Serial][sig-compute]Infrastructure"+text, Serial, decorators.SigCompute, args)
+	return Describe("[Serial][sig-compute]Infrastructure "+text, Serial, decorators.SigCompute, args)
```
Comment (fossedihelm) [inline tests/infrastructure/infrastructure.go:29]: It is out of the scope of this PR, but I really don't like hiding the `Serial` decorator behind a custom name.

**Review by fossedihelm — COMMENTED** [review]
Thank you for this wonderful work @jean-edouard 
/lgtm
/hold
Some small comments below, feel free to unhold if you want to address them in a separate PR!

File: pkg/virt-handler/heartbeat/ksm.go [hunk 3]
```diff
@@ -0,0 +1,299 @@
...
+	total, available, err := getTotalAndAvailableMem()
+	if err != nil {
+		return ksm, err
+	}
+	ksm.pages, err = getKsmPages()
+	if err != nil {
+		return ksm, err
+	}
+
+	ksm.sleep = sleepMsBaseline * (16 * 1024 * 1024) / (total - available)
```
Comment (acardace) [inline pkg/virt-handler/heartbeat/ksm.go:136]: Can you explain the `ksm.sleep` calculations? A comment would be great.

File: pkg/virt-handler/heartbeat/heartbeat.go [hunk 4]
```diff
@@ -135,10 +135,20 @@ func (h *HeartBeat) do() {
 	if h.clusterConfig.CPUManagerEnabled() {
 		cpuManagerEnabled = h.isCPUManagerEnabled(h.cpuManagerPaths)
 	}
-	data = []byte(fmt.Sprintf(`{"metadata": { "labels": {"%s": "%s", "%s": "%t"}, "annotations": {"%s": %s}}}`,
+
+	node, err := h.clientset.Nodes().Get(context.Background(), h.host, metav1.GetOptions{})
```
[the remark this answers is not in the document]
Comment (fossedihelm) [reply pkg/virt-handler/heartbeat/heartbeat.go:139]: FYI In general, adding informer in node level components should be avoided :) https://github.com/kubevirt/kubevirt/blob/main/docs/reviewer-guide.md#common-architecture-flaws-to-avoid

File: tests/infrastructure/virt-handler.go (hunk 1 above)
Comment (jean-edouard) [reply tests/infrastructure/virt-handler.go:109]: Agreed

File: tests/infrastructure/infrastructure.go (hunk 2 above)
Comment (jean-edouard) [reply tests/infrastructure/infrastructure.go:29]: Agreed

File: pkg/virt-handler/heartbeat/ksm.go (hunk 3 above)
Comment (jean-edouard) [reply pkg/virt-handler/heartbeat/ksm.go:139]: Comment added

File: pkg/virt-handler/heartbeat/ksm.go [hunk 5]
```diff
@@ -13,85 +36,264 @@ import (
...
+	if ok {
+		value, err := strconv.Atoi(override)
+		if err == nil {
+			return value
+		} else {
+			log.DefaultLogger().Reason(err).Errorf("failed to parse %s override value, using default", param)
+		}
+	}
+
+	return defaultValue
```
[the remark this answers is not in the document]
Comment (jean-edouard) [reply pkg/virt-handler/heartbeat/ksm.go:230]: Added a big unit test with a bunch of overrides, PTAL!

**Comment by fossedihelm** [comment]
/lgtm
Thanks @jean-edouard

**Comment by fabiand** [comment]
Wow.

**Comment by acardace** [comment]
> Wow.

Lots of goodies in 1.1 ;)

File: pkg/virt-handler/heartbeat/heartbeat.go (hunk 4 above)
Comment (xpivarc) [inline pkg/virt-handler/heartbeat/heartbeat.go:139]: @jean-edouard Why do we need this?

**Pull request merged**